Use writeHead instead of setHeader/getHeader.

mcollina opened the pull request:
This gives us a nice +10% on throughput.

Before:

```
$ autocannon -c 100 -d 5 -p 10 localhost:3000
Running 5s test @ http://localhost:3000
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 5.34    16.27   189.41
Req/Sec      18340.8 195.81  18691
Bytes/Sec    2.71 MB 52.4 kB 2.78 MB

92k requests in 5s, 13.7 MB read
```

After:

```
$ autocannon -c 100 -d 5 -p 10 localhost:3000
Running 5s test @ http://localhost:3000
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 4.86    15.12   264.2
Req/Sec      20155.2 294.93  20661
Bytes/Sec    3 MB    64.2 kB 3.08 MB

101k requests in 5s, 15 MB read
```

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

allevo commented on examples/example.js:
```diff
@@ -19,7 +19,6 @@ const opts = {
 
 fastify
   .get('/', opts, function (req, reply) {
-    reply.header('Content-Type', 'application/json').code(200)
```
Why this line is removed?

mcollina replied:
it was not needed, those are the defaults.

mcollina commented:
The trampoline is added by V8 when there is an arity mishmatch between how you call a function and how is defined. This normally has no influence on performance, unless when it does.

nwoltman commented:
Fixing the performance of calling functions with a different arity is on the V8 roadmap if anyone's interested in following along: https://bugs.chromium.org/p/v8/issues/detail?id=7460

nwoltman commented on lib/handleRequest.js:
```diff
@@ -13,8 +13,10 @@ const inputSchemaError = fastJsonStringify(schemas.inputSchemaError)
+  var Request = context.Request
+  var Reply = context.Reply
+  var request = new Request(params, req, urlUtil.parse(req.url, true).query, headers, req.log)
+  var reply = new Reply(res, context, request)
```
Does the old code prevent V8 from optimizing the constructor call?

mcollina replied:
not really, but I had to fix a bad bug in my change and I forgot these in. I can revert this bit if you want.

nwoltman commented on lib/reply.js:
```diff
@@ -47,20 +48,20 @@ Reply.prototype.send = function (payload) {
   }
 
-  const contentType = this.res.getHeader('Content-Type')
+  const contentType = this._headers['content-type']
```
The `Content-Type` could have been set on the `res` object by a hook or middleware. I am pretty worried about the potential complications of having response headers stored in 2 different places.

delvedor replied:
This can be easily solved with a `reply.getHeader` that fallbacks to the basic response headers.
I tested this in my local copy and it works perfectly.
If @mcollina is ok I can send the change directly in this branch.

nwoltman commented on lib/reply.js:
```diff
@@ -140,27 +141,36 @@ function wrapOnSendEnd (err, reply, payload) {
 
-  if (!reply.res.getHeader('Content-Length')) {
-    reply.res.setHeader('Content-Length', '' + Buffer.byteLength(payload))
+  if (!reply._headers['content-length']) {
```
Same thing as my other comment about the header possibly being set on `res`.

delvedor replied:
This is the final handler of the `onSend` hooks, so if a user wants to alter the headers it will use the `reply` interface. I hope that no one sets the `content-length` header inside a `onRequest` or `preHandler` hook.

nwoltman replied:
I also hope that no one will do that, but we can't stop them from doing it so it's a case that needs to be handled.

delvedor replied:
I don't agree, I don't think we should drop a considerable amount of ops/sec just because a user **may** do something that is wrong.

mcollina commented:
I've removed `reply.getHeader` as I don't think it's needed as a public method and when it's needed we have the functionality to add it. I prefer not to add new API methods for performance updates if it's possible.

I've added a fast-path to avoid performance issues for Node 8+.

delvedor commented:
On my machine this change give us *1500/2000 req/sec* more! :rocket:

jsumners commented on lib/reply.js:
```diff
@@ -47,20 +50,20 @@ Reply.prototype.send = function (payload) {
-  const contentType = this.res.getHeader('Content-Type')
-  const contentTypeNotSet = contentType === undefined
+  var contentType = getHeader(this, 'content-type')
+  var contentTypeNotSet = contentType === undefined
```
Is the change to `var` here an optimization?

mcollina replied:
It might be, but I didn't measure without.

jsumners replied:
Maybe change it back for consistency?

delvedor replied:
@jsumners as far as I know `var` is slightly faster than `const` because of the scoping.

mcollina replied:
@jsumners @delvedor I tried reverting the const -> var change and the context thing, and I got a slight downgrade of the throughput. I prefer to leave them as they are in this PR.

jsumners replied:
Which means the answer to the question is "yes".

mcollina replied:
yes.

mcollina commented:
@nwoltman @allevo are you ok with these?

nwoltman commented on lib/reply.js:
```diff
@@ -292,5 +310,36 @@ function notFound (reply) {
+  key = key.toLowerCase()
+  var res = reply.res
+  var value = reply._headers[key]
+  if (value === undefined && res.hasHeader(key)) {
```
Using `res.hasHeader` doesn't seem like it would help much. It's almost the exact same code as `res.getHeader` so it would be simpler just to have one function that's the same as `getHeaderFallback`.

mcollina replied:
I do not have investigated why, but `res.getHeader` slow us down, while `res.hasHeader` does not.

nwoltman replied:
Oh that is very interesting.

nwoltman commented on lib/reply.js:
```diff
@@ -140,27 +143,36 @@ function wrapOnSendEnd (err, reply, payload) {
   if (payload === undefined || payload === null) {
     reply.sent = true
-    reply.res.end()
+    reply._headers['content-length'] = '0'
```
The `Content-Length` header should not be set here. There's rules about it in [RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.3.2).

nwoltman commented:
@delvedor's idea about having a public `getHeader` method was correct. It is necessary. For example, a compression plugin needs to know the `Content-Type` to determine whether or not it should compress the payload.

nwoltman commented on lib/reply.js:
```diff
@@ -140,27 +143,36 @@ function wrapOnSendEnd (err, reply, payload) {
   reply.sent = true
-  reply.res.end(payload)
+
+  res.writeHead(res.statusCode, reply._headers)
```
You can avoid the ArgumentsAdaptorTrampoline in `res.writeHead` calls too by passing in a third argument.
https://github.com/nodejs/node/blob/master/lib/_http_server.js#L200

mcollina replied:
Unfortunately if I pass `null` or `undefined` all tests fails.

nwoltman replied:
Oh that's too bad. I guess maybe the `writeHead` code is different for older versions of Node (it looks like it should work with the current code on master).

allevo replied:
Have you tried `res.writeHead(res.statusCode, null, reply._headers)`?
Or `res.writeHead(res.statusCode, http.STATUS_CODE[res.statusCode], reply._headers)`?

mcollina replied:
@allevo can't make it to work, but I would like to land this before the 1st of March, and this can be worked out afterwards.

allevo replied:
There's another problem. Using the three parameters version on http2 a warning is emitted:
`(node:2103) UnsupportedWarning: Status message is not supported by HTTP/2 (RFC7540 8.1.2.4)`

mcollina commented:
Having  `header()` to _set_ and  `getHeader` to _get_ is very confusing. We need a better name.

mcollina commented:
@nwoltman updated to not set `content-length: 0`.

mcollina commented:
An option might be to remove `headers(headersToSet)` and just expose `_headers` as `headers`.
We can also name it as `headersToSend`.

nwoltman commented:
@mcollina I agree having `header()` and `getHeader()` would be confusing. We could add `setHeader()` and remove `header()` in a later version.

allevo commented on lib/reply.js:
```diff
@@ -105,7 +108,7 @@ Reply.prototype.serializer = function (fn) {
 
 Reply.prototype.type = function (type) {
-  this.res.setHeader('Content-Type', type)
+  this._headers['content-type'] = type
```
`this.header` ?

mcollina replied:
why going for a slow-path when there is a fast-path?

allevo commented on lib/reply.js:
```diff
@@ -196,11 +207,15 @@ function sendStream (payload, res, reply) {
 
+  for (var key in reply._headers) {
+    res.setHeader(key, reply._headers[key])
+  }
```
Why not `res.writeHead`?

mcollina replied:
I’ll add a comment. We want to return 500s if the streams error before trying to send any data. That is the only way.

allevo replied:
Good point. I'm 👍  for the comment

allevo commented on lib/reply.js:
```diff
@@ -292,5 +309,36 @@ function notFound (reply) {
+// ponyfill for hasHeader. It has been intoroduced into Node 7.7,
+// so it's ok to use it in 8+
+{
+  const v = process.version.match(/v(\d)/)[1]
```
`process.version.match(/v(\d+)/)[1]` (+ added)

allevo commented on test/get.test.js:
```diff
@@ -362,7 +362,7 @@ fastify.listen(0, err => {
       t.error(err)
       t.strictEqual(response.statusCode, 200)
-      t.strictEqual(response.headers['content-length'], '0')
+      t.strictEqual(response.headers['content-length'], undefined)
```
IMHO this is wrong.
The content length should always be set

jsumners replied:
The spec allows for it to *not* be set -- https://tools.ietf.org/html/rfc2616#section-14.13

allevo replied:
yeah and in the rfc:

> Applications SHOULD use this field to indicate the transfer-length of the message-body

where "SHOULD" means:

> This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course

(https://tools.ietf.org/html/rfc2119)

jsumners replied:
Yes, but it isn't a "MUST".

allevo commented on test/internals/reply.test.js:
```diff
@@ -193,7 +194,7 @@ test('within an instance', t => {
       http.get(url, (response) => {
         t.strictEqual(response.headers['x-onsend'], 'yes')
-        t.strictEqual(response.headers['content-length'], '0')
+        t.strictEqual(response.headers['content-length'], undefined)
```
here too

mcollina commented:
@allevo @jsumners take a look, I have implemented things according to the spec. I'm not sure how tests were passing before.

nwoltman commented on lib/reply.js:
```diff
@@ -140,27 +143,44 @@ function wrapOnSendEnd (err, reply, payload) {
+    // we cannot send a content-length for 304 and 204, and all status code
+    // < 200.
+    if (statusCode >= 200 && statusCode !== 204 && statusCode !== 304) {
+      reply._headers['content-length'] = '0'
```
> A server MAY send a Content-Length header field in a response to a HEAD request (Section 4.3.2 of [RFC7231]); a server MUST NOT send Content-Length in such a response unless its field-value equals the decimal number of octets that would have been sent in the payload body of a response if the same request had used the GET method.

If a user sends a 200 response to a HEAD request, this will most likely be the wrong Content-Length.

jsumners replied:
Except this is within the condition that the payload is either `undefined` or `null`. Thus, there isn't a content length.

mcollina replied:
what should it be in that case? Should we check if there is a content-length set before setting it to zero?

mcollina replied:
I’m completely lost. This code passes all our tests in master without changes.

jsumners replied:
@mcollina I believe you are correct with the code as it is. Look at the full context:

https://github.com/fastify/fastify/blob/578517d46d9edd90d11f157c5169113796019f0d/lib/reply.js#L149-L163

The payload size should always be 0 where you are explicitly setting the content length because a payload does not exist.

nwoltman replied:
@jsumners If it's a HEAD request, the user may decide to only set the headers and not send the payload itself (even though it would be stripped by Node). Users can do this to avoid spending time generating a payload.

@mcollina In that case the Content-Length should not be set if it was not already set by a user.

nwoltman replied:
It's okay. I'm not going to block this.

jsumners replied:
It's still appropriate to send `Content-length: 0` in that case because it is true. Headers do not count.
